src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+  --> src/apply.rs:72:13
+   |
+72 |             self.notifier,
+   |             ^^^^^^^^^^^^^ move occurs because `self.notifier` has type `Notifier`, which does not implement the `Copy` trait
+```
+
+After discussing with [Barbara][], Alan adds `Arc` to provide a shared ownership between async tasks:
+
+```rust
+use async_std::{sync::Arc, task};
+
+pub struct Applier {
+    notifier: Arc<notify::Notifier>,
+    ...
+}
+
+impl Applier {
+    pub fn new(req_timeout: Duration) -> Self {
+        ...
+        Self {
+            ...
+            notifier: Arc::new(notify::Notifier::new()),
+            ...
+        }
+    }
+    ...
+
+    pub async fn start(&self) -> io::Result<()> {
+        task::spawn(apply_async(
+            self.notifier.clone(),
+            ...
+        ));
+        ...
+        Ok(())
+    }
+    ...
+
+
+pub async fn apply_async(
+    notifier: Arc<notify::Notifier>,
+    ...
+) -> io::Result<()> {
+  ...
+```
+
+Alan is satisfied with the compilation success for the moment, but doesn't feel confident about the production readiness of Rust async.
+
+### Implementing HTTP server handler
+
+Familiar with Go standard libraries, Alan implemented the following request handler without any third-party dependencies:
+
+```go
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+type Handler interface {
+	start()
+}
+
+type handler struct {
+	listenerPort uint64
+	applier      Applier
+}
+
+func (hd *handler) start() {
+	hd.applier.start()
+
+	serverMux := http.NewServeMux()
+	serverMux.HandleFunc("/echo", hd.wrapFunc(handleRequest))
+
+	httpServer := &http.Server{
+		Addr:    fmt.Sprintf(":%d", hd.listenerPort),
+		Handler: serverMux,
+	}
+
+	tch := make(chan os.Signal, 1)
+	signal.Notify(tch, syscall.SIGINT)
+	done := make(chan struct{})
+	go func() {
+		httpServer.Close()
+		close(done)
+	}()
+
+	if err := httpServer.ListenAndServe(); err != nil {
+		fmt.Printf("http server error: %v\n", err)
+	}
+	select {
+	case <-done:
+	default:
+	}
+
+	if err := hd.applier.stop(); err != nil {
+		panic(err)
+	}
+}
+
+func (hd *handler) wrapFunc(fn func(applier Applier, w http.ResponseWriter, req *http.Request)) func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		fn(hd.applier, w, req)
+	}
+}
+
+func handleRequest(applier Applier, w http.ResponseWriter, req *http.Request) {
+	switch req.Method {
+	case "POST":
+		var echoRequest EchoRequest
+		err := json.NewDecoder(req.Body).Decode(&echoRequest)
+		if err != nil {
+			fmt.Fprintf(w, "failed to read request %v", err)
+			return
+		}
+		s, err := applier.apply(Request{echoRequest: &echoRequest})
+		if err != nil {
+			fmt.Fprintf(w, "failed to apply request %v", err)
+			return
+		}
+		fmt.Fprint(w, s)
+
+	default:
+		http.Error(w, "Method Not Allowed", 405)
+	}
+}
+```
+
+For Rust, Alan is faced with multiple choices again:
```

Comment:
In this case I think rust is better here. I think web server handling is too complex to have a "standard" way of doing things. Having choices is better than being forced with a single choice.

Reply:
Please remember that these stories are meant to reflect lived experiences. It's not necessarily relevant whether having choice is better than having a standard way of doing things. If this caused Alan confusion or hesitation, it should be noted.